egs/librispeech/ASR/transducer_emformer/streaming_decode.py

```diff
@@ -147,126 +147,196 @@ def get_parser():
+        feature_lens = torch.tensor([features.size(1)], device=device)
+        (
+            encoder_out,
+            encoder_out_lens,
+            stream.states,
+        ) = model.encoder.streaming_forward(
+            features,
+            feature_lens,
+            stream.states,
+        )
+        greedy_search(
+            model=model,
+            stream=stream,
+            encoder_out=encoder_out[0],
+            sp=sp,
+        )
 
-    hyp = [blank_id] * context_size
-
-    decoder_input = torch.tensor(hyp, device=device, dtype=torch.int64).reshape(
-        1, context_size
-    )
+    if stream.feature_extractor.is_last_frame(stream.num_fetched_frames - 1):
+        assert len(stream.feature_frames) < chunk_length
+
+        if len(stream.feature_frames) > 0:
+            this_chunk = stream.feature_frames[:chunk_length]
+            stream.feature_frames = []
+            features = torch.cat(this_chunk, dim=0)  # (T, feature_dim)
+            features = features.to(device).unsqueeze(0)  # (1, T, feature_dim)
+            features = torch.nn.functional.pad(
+                features,
+                (0, 0, 0, chunk_length - features.size(1)),
+                value=LOG_EPSILON,
+            )
+            feature_lens = torch.tensor([features.size(1)], device=device)
+            (
+                encoder_out,
+                encoder_out_lens,
+                stream.states,
+            ) = model.encoder.streaming_forward(
+                features,
+                feature_lens,
+                stream.states,
+            )
+            greedy_search(
+                model=model,
+                stream=stream,
+                encoder_out=encoder_out[0],
+                sp=sp,
+            )
 
-    decoder_out = model.decoder(decoder_input, need_pad=False)
 
+def decode_one_utterance(
+    audio_samples: torch.Tensor,
+    model: nn.Module,
+    stream: Stream,
+    params: AttributeDict,
+    sp: spm.SentencePieceProcessor,
+):
+    """Decode one utterance.
+    Args:
+      audio_samples:
+        A 1-D float32 tensor of shape (num_samples,) containing the
+        audio samples.
+      model:
+        The RNN-T model.
+      feature_extractor:
+        The feature extractor.
+      params:
+        It is the return value of :func:`get_params`.
+      sp:
+        The BPE model.
+    """
     i = 0
     num_samples = audio_samples.size(0)
     while i < num_samples:
-        # Note: The current approach of computing the features is not ideal
-        # since it re-computes the features for the right context.
-        chunk = audio_samples[i : i + chunk_size]  # noqa
-        i += segment_length
-        if chunk.size(0) < chunk_size:
-            chunk = torch.nn.functional.pad(
-                chunk, pad=(0, chunk_size - chunk.size(0))
-            )
-        features = fbank(chunk)
-        feature_lens = torch.tensor([features.size(0)], device=params.device)
+        # Simulate streaming.
```

Comment:
The code here shows how the online feature extractor works.

We can send an arbitrary number of samples to the extractor. There is a buffer in the extractor recording the samples that have not been used in computing features.